spec/text-editor-spec.coffee

```diff
@@ -5924,3 +5927,58 @@ describe "TextEditor", ->
   describe "::getElement", ->
     it "returns an element", ->
       expect(editor.getElement() instanceof HTMLElement).toBe(true)
+
+  describe "verticalScrollMargin", ->
+    it "scrolls the window down before reaching the bottom with a given margin", ->
+      # give plenty of text to allow for scrolling
+      editor.setText('a\n'.repeat(50))
+
+      # make sure editor element has valid dimensions
+      element = editor.getElement()
+      element.style.width  = '200px'
+      element.style.height = '120px'
+      jasmine.attachToDOM(element)
+
+      component = element.getComponent()
+      # for a height of 120px and a line height of 17px, this will be 7 rows
+      rowCount = component.getLastVisibleRow()
+
+      for scrollMargin in [0..4]
+        editor.update({verticalScrollMargin: scrollMargin})
+
+        # margin is clamped so that it is no larger than half the screen
+        maxScrollMargin = Math.floor(((component.getScrollContainerClientHeight() / editor.getLineHeightInPixels()) - 1) / 2)
```

Comment:
Sure thing. Could you elaborate a little on what conditions you think are best as separate tests? There's only one actual expect().toBe in each test, which tests if the top/bottom of the window is where it is expected given the size of the vertical scroll margin. I'm not sure how I could test the clamping separately, since the test would fail if it weren't present.

Do you maybe mean separately testing the different settings of scrollMargin and row positions in the for loops? So e.g., a hard coded test for scrolling down to line 6, with a margin size of 3, and a window size of 7 (so the bottom would be expected to be line 9). In this situation, it would be possible to test clamping only in scenarios where the clamping is necessary. I thought it would be a more exhaustive test to iterate over a range of values and handle under/overflow for those cases that it does happen, but if you think it would be clearer doing individual tests that works too. The scenario where clamping is required will still require this more complicated clamping test in order to be a correct test of the functionality though.

Or maybe you had something else in mind? Let me know and I will gladly update the tests.